### DISCUSSION TITLE
Possible to use conditional in the "env" section of a job?

### DISCUSSION THREAD

**Discussion by bagel-dawg** [community/community#25725] [Actions]
<p>I have several use cases where the environment variables for a job change depending on if I am on the master branch or not, or for example if the build up to the point has failed.</p>
<p>My code for sending a slack message based on build status looks something like this:</p>
<pre><code class="lang-auto">- name: Success Send Build Notification
  uses: rtCamp/action-slack-notify@v2.1.0
  if: github.ref == 'refs/heads/master' &amp;&amp;  success()
  env:
    SLACK_COLOR: "good"
    SLACK_USERNAME: "Github Actions"
    SLACK_ICON_EMOJI: ":ocotcat:"
    SLACK_TITLE: "This is a GitHub Actions build!"
    SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
    SLACK_CHANNEL: "#channel"
    SLACK_MESSAGE: "This build has passed!."

- name: Failed Send Build Notification
  uses: rtCamp/action-slack-notify@v2.1.0
  if: github.ref == 'refs/heads/master' &amp;&amp;  failure()
  env:
    SLACK_COLOR: "danger"
    SLACK_USERNAME: "Github Actions"
    SLACK_ICON_EMOJI: ":ocotcat:"
    SLACK_TITLE: "This is a GitHub Actions build!"
    SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
    SLACK_CHANNEL: "#channel"
    SLACK_MESSAGE: "This build has failed. :("
</code></pre>
<p>It would be much more convenient if I could instead combine this into a single code block which conditional env vars, something like this?</p>
<pre><code class="lang-auto">- name: Send Build Notification
  uses: rtCamp/action-slack-notify@v2.1.0
  if: github.ref == 'refs/heads/master' &amp;&amp;  success()
  env:
    SLACK_COLOR: "${{ job.status }} === 'success' ? 'good' : 'danger'"
    SLACK_USERNAME: "Github Actions"
    SLACK_ICON_EMOJI: ":ocotcat:"
    SLACK_TITLE: "This is a GitHub Actions build!"
    SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
    SLACK_CHANNEL: "#channel"
    SLACK_MESSAGE: "This job has ${{ job.status }} === 'success' ? 'passed!' : 'failed. :('"
</code></pre>
<p>Is there a way to do this, or perhaps another creative, compact way of accomplishing the same goal?</p>

**Comment by Simran-B** [comment, marked as answer]
<p>There is no support for a ternary operator, but there are workarounds. You can try the following:</p>
<pre><code class="lang-auto">- name: Send Build Notification
  uses: rtCamp/action-slack-notify@v2.1.0
  if: github.ref == 'refs/heads/master' &amp;&amp;  success()
  env:
    SLACK_COLOR: "${{ job.status == 'success' &amp;&amp; 'good' || 'danger' }}"
    SLACK_USERNAME: "Github Actions"
    SLACK_ICON_EMOJI: ":ocotcat:"
    SLACK_TITLE: "This is a GitHub Actions build!"
    SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
    SLACK_CHANNEL: "#channel"
    SLACK_MESSAGE: "This job has ${{ job.status == 'success' &amp;&amp; 'passed!' || 'failed. :(' }}"
</code></pre>
<p>A limitation here is that the value after <code>&amp;&amp;</code> must not be falsey, because in that case the <code>||</code> (or)  branch would be taken incorrectly. Not a problem here however, because it’s not an empty string or the like.</p>
<p>Another pseudo ternary expression that I haven’t tested yet, but which should work in principle and doesn’t suffer from above limitation:</p>
<pre><code>fromJSON('["danger", "good"]')[job.status == 'success']
</code></pre>
<p>This relies on type coercion, an implicit cast from boolean <code>true</code> to <code>1</code> or <code>false</code> to <code>0</code>, which is then used as array index. As array literals aren’t supported (yet), <code>fromJSON()</code> is required to construct the array.</p>

**Reply by devSajan** [reply]
when creating this variable it's not working
FILE_PATH: "${{ env.ACT &&  '$HOME/file_path' || '/home/file_path' }}"
I am getting FILE_PATH value as $HOME/file_path instead of /home/user/file_path
how can i do that?

**Comment by BrightRan** [comment]
<p><a class="mention" href="/bagel-dawg">@bagel-dawg</a>,</p>
<blockquote>
<p>Possible to use conditional in the “env” section of a job?</p>
</blockquote>
<p>This is not supported currently.</p>
<p>As a workaround, you can try to configure your workflow like as below:</p>
<pre><code class="lang-yaml">jobs:
  build:
    . . .

  send_slack_msg:
    needs: [build]
    if: ${{ github.ref == 'refs/heads/master' &amp;&amp; always() }}
    runs-on: ubuntu-latest
    steps:
      - name: Set environment variables
        run: |
          echo "SLACK_USERNAME=Github Actions" &gt;&gt; $GITHUB_ENV
          echo "SLACK_ICON_EMOJI=:ocotcat:" &gt;&gt; $GITHUB_ENV
          echo "SLACK_TITLE=This is a GitHub Actions build!" &gt;&gt; $GITHUB_ENV
          echo "SLACK_WEBHOOK=${{ secrets.SLACK_WEBHOOK }}" &gt;&gt; $GITHUB_ENV
          echo "SLACK_CHANNEL=#channel" &gt;&gt; $GITHUB_ENV

          if [[ ${{ needs.build.result }} == success ]]; then
            echo "SLACK_COLOR=good" &gt;&gt; $GITHUB_ENV
            echo "SLACK_MESSAGE=This build has passed!" &gt;&gt; $GITHUB_ENV
          else
            echo "SLACK_COLOR=danger" &gt;&gt; $GITHUB_ENV
            echo "SLACK_MESSAGE=This build has failed." &gt;&gt; $GITHUB_ENV
          fi

      - name: Send Build Notification
        uses: rtCamp/action-slack-notify@v2.1.0
</code></pre>

**Comment by bagel-dawg** [comment]
<p>Thanks! I’ll try out both solutions and select one.</p>

**Comment by ansiwen** [comment]
<p>I would recommend <code>fromJSON('{"false": "danger", "true": "good"}')[job.status == 'success']</code> because it is more explicit and is a clean solution. With this pattern you can also do more sophisticated mapping like <code>fromJSON('{"Linux":"so","macOS":"dylib","Windows":"dll"}')[runner.os]</code>.</p>

**Reply by m4heshd** [reply]
For anybody who's trying to use this to set an env var, the `runner` context is not available for `env`.

**Reply by ansiwen** [reply]
> For anybody who's trying to use this to set an env var, the `runner` context is not available for `env`.

Why would you think that? I'm using it to set an env variable: https://github.com/Nitrokey/nethsm-pkcs11/blob/b43302dfc0352a07f86de2869c35bdd12b4d1944/.github/workflows/go.yml#L28

**Reply by m4heshd** [reply]
> Why would you think that? I'm using it to set an env variable: https://github.com/Nitrokey/nethsm-pkcs11/blob/b43302dfc0352a07f86de2869c35bdd12b4d1944/.github/workflows/go.yml#L28

I'm not sure what kind of magic you're using there but [here's the documentation](https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability). Check `env`. Only `github, secrets, inputs` contexts are available for `env`. So it working just for you is pretty magical.

Here's the error I ended up with, when I tried to use `runner` context.

![image](https://user-images.githubusercontent.com/6711514/197361026-8cf2f277-1a53-4551-872d-5ecf8710ac6c.png)

Yes, I tried to use it exactly how you're utilizing it, in a job. Then I changed it to `inputs` _(because I actually have inputs on my workflow)_ and it immediately worked. Didn't change anything else. 

One difference I'm seeing in your case is the matrix strategy. But I don't need that if I'm using a matrix. I can simply use `include` which I use all the time. BTW you can too. Yes, it's not a one liner but pretty straight forward and provides much better readability.

Simply do:

```yaml
  build:
    name: Build on ${{ matrix.os }}
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        os: [ubuntu-latest, windows-latest, macOS-latest]
        include:
          - os: ubuntu-latest
            lib_suffix: so
          - os: windows-latest
            lib_suffix: dll
          - os: macOS-latest
            lib_suffix: dylib
    steps:
    - uses: actions/checkout@v2

    - name: Set up Go
      uses: actions/setup-go@v2
      with:
        go-version: 1.17

    - name: Build
      run: go build -o out/p11nethsm.${{ matrix.lib_suffix }} -buildmode=c-shared -ldflags=-w -v
```

Or this:

```yaml
include:
  - {os: ubuntu-latest, lib_suffix: so}
  - {os: windows-latest, lib_suffix: dll}
  - {os: macOS-latest, lib_suffix: dylib}
```

**Reply by m4heshd** [reply]
**UPDATE:**

Ok I just realized, `runner` is available for `jobs.<job_id>.steps.env` which you're utilizing. It's just not available for job's top level `env`. That's pretty odd since the runner is already initialized prior to starting a job. Oh well.. 🤷🏼

**Reply by ansiwen** [reply]
Oh, I see. No, that cannot work, because the runner is part of the job context. Actually it's even itself an environment variable of the job context, so that would be a recursion: [Detecting the operating system](https://docs.github.com/en/actions/learn-github-actions/environment-variables#detecting-the-operating-system)

**Reply by m4heshd** [reply]
I know, that's why I said "That's pretty odd". It's not like a new runner is spawned for each step. It feels like it should clearly be available for job's `env`. That can be the last task executed just before running the first step. It makes perfect sense since that `env` is not available for any other top level key. Just for the steps and jobs that `needs` this particular job. It's just like setting an output on the very first step.
